src/main/scala/beam/agentsim/agents/vehicles/VehicleEnergy.scala

```diff
@@ -0,0 +1,209 @@
+                case None => numberOfLanesFilter += numberOfLanesBin -> rate
+              }
+            }
+            case None => gradePercentFilter += gradePercentBin -> mutable.Map(numberOfLanesBin -> rate)
+          }
+        }
+        case None =>
+          currentRateFilter += speedInMilesPerHourBin ->
+            mutable.Map(gradePercentBin -> mutable.Map(numberOfLanesBin -> rate))
+      }
+    })
+    currentRateFilter.toMap.map {
+      case (speedInMilesPerHourBin, gradePercentMap) =>
+        speedInMilesPerHourBin -> gradePercentMap.toMap.map {
+          case (gradePercentBin, lanesMap) => gradePercentBin -> lanesMap.toMap
+        }
+    }
+  }
+
+  private def convertRecordStringToRange(recordString: String) = {
+    Range(recordString.replace(",", ":").replace(" ", ""))
+  }
+}
+
+class VehicleEnergy(
+  consumptionRateFilterStore: ConsumptionRateFilterStore,
+  linkToGradeRecordsIterableUsing: CsvParser => Iterable[Record]
+) {
+  private lazy val log = LoggerFactory.getLogger(this.getClass)
+
+  private val settings = new CsvParserSettings()
+  settings.detectFormatAutomatically()
+  private val csvParser = new CsvParser(settings)
+
+  type ConsumptionRateFilter = Map[Range, Map[Range, Map[Range, Double]]] //speed->(gradePercent->(numberOfLanes->rate))
+  private lazy val linkIdToGradePercentMap = loadLinkIdToGradeMapFromCSV
+  val conversionRateForJoulesPerMeterConversionFromGallonsPer100Miles = 746.86
+
+  def getFuelConsumptionEnergyInJoulesUsing(
+    fuelConsumptionDatas: IndexedSeq[BeamVehicle.FuelConsumptionData],
+    overallFallBack: IndexedSeq[BeamVehicle.FuelConsumptionData] => Double,
+    singleFallBack: BeamVehicle.FuelConsumptionData => Double
+  ): Double = {
+    val consumptionsInJoules: IndexedSeq[Double] = fuelConsumptionDatas
+      .flatMap(fuelConsumptionData => {
+        for {
+          rateInJoulesPerMeter <- getRateUsing(fuelConsumptionData, singleFallBack)
+          distance             <- fuelConsumptionData.linkLength
+        } yield rateInJoulesPerMeter * distance
+      })
+    if (consumptionsInJoules.nonEmpty) consumptionsInJoules.sum else overallFallBack(fuelConsumptionDatas)
+  }
+
+  private def getRateUsing(fuelConsumptionData: BeamVehicle.FuelConsumptionData,
+                           fallBack: BeamVehicle.FuelConsumptionData => Double): Option[Double] = {
+    val BeamVehicle.FuelConsumptionData(linkId, vehicleType, numberOfLanesOption, _, _, _, speedInMilesPerHourOption, _, _, _) =
+      fuelConsumptionData
+    val numberOfLanes: Int = numberOfLanesOption.getOrElse(0)
+    val speedInMilesPerHour: Double = speedInMilesPerHourOption.getOrElse(0)
+    val gradePercent: Double = linkIdToGradePercentMap.getOrElse(linkId, 0)
+    consumptionRateFilterStore.getPrimaryConsumptionRateFilterFor(vehicleType).flatMap(
+      consumptionRateFilterFuture => getRateUsing(consumptionRateFilterFuture,
+        numberOfLanes, speedInMilesPerHour, gradePercent)
+    ).orElse(
```

Comment:
Secondary energy consumption is not a backup if there's no match on primary. Totally different meaning. 

It's intended for 2 fuel systems. I.e. a plug-in hybrid (like a Volt) you charge up the battery and cruise on all-electric for first 50 miles, then when the battery runs down, it starts using gasoline and runs like a regular hybrid.

So we only want to use secondary if the fuel tank in primary is out.

But let's not solve that problem here in this ticket, let's get things working and solve that in a separate issue.

Reply:
Ahhh, I didn't realize that - although it makes sense. Yah, I'll remove secondary for right now and create an issue